File: ktkernel/subsystems/vmm/vmm.cpp
#include "subsystems/vmm/vmm.h"
#include "limine/requests.h"
#include "mem/memory.h"
#include "systems.h"

#include <kt/intrin.h>

namespace kt_kernel
{
    void vmm::initialize()
    {
        hhdm_offset = limine::hhdm_request.response->offset;
        pml4_physical = read_cr3() & static_cast<uint64_t>(page_flags::address_mask);
    }

    uint64_t* vmm::get_pte(uint64_t virt, bool create)
    {
        uint64_t* table = reinterpret_cast<uint64_t*>(pml4_physical + hhdm_offset);

        for (int level = 3; level > 0; level--) {
            uint64_t index = (virt >> (12 + 9 * level)) & 0x1FF;

            if (!(table[index] & static_cast<uint64_t>(page_flags::present))) {
                if (!create)
                    return nullptr;

                uint64_t frame = g_kernel_context->pmm->allocate_frame();
                if (frame == 0)
                    return nullptr;

                auto* page = reinterpret_cast<uint64_t*>(frame + hhdm_offset);
                memset(page, 0, 512 * sizeof(uint64_t));

                table[index] = frame | static_cast<uint64_t>(page_flags::present | page_flags::writable);
            }
            table = reinterpret_cast<uint64_t*>((table[index] & static_cast<uint64_t>(page_flags::address_mask)) +
                                                 hhdm_offset);
        }

        return &table[(virt >> 12) & 0x1FF];
    }

    bool vmm::map(uint64_t virt, uint64_t phys, uint64_t flags)
    {
        uint64_t* pte = get_pte(virt, true);
        if (!pte)
            return false;

        *pte = (phys & static_cast<uint64_t>(page_flags::address_mask)) | flags |
               static_cast<uint64_t>(page_flags::present);
        return true;
    }

    void vmm::unmap(uint64_t virt)
    {
        uint64_t* pte = get_pte(virt, false);
        if (!pte)
            return;

        *pte = 0;
        invlpg(virt);
    }

    uint64_t vmm::virtual_to_physical(uint64_t virt)
    {
        uint64_t* pte = get_pte(virt, false);
        if (!pte || !(*pte & static_cast<uint64_t>(page_flags::present)))
            return 0;

        return (*pte & static_cast<uint64_t>(page_flags::address_mask)) | (virt & 0x1FF);
    }

    uint64_t vmm::physical_to_virtual(uint64_t phys)
    {
        return phys + hhdm_offset;
    }

    void vmm::activate()
    {
        write_cr3(pml4_physical);
    }
} // namespace kt_kernel

File: ktkernel/src/memory/vmm.rs
use spin::Mutex;

use super::pmm::PMM;
use crate::limine::HHDM;
use crate::utils::intrin::{invlpg, read_cr3, write_cr3};

pub mod flags {
    pub const PRESENT: u64 = 1 << 0;
    pub const WRITABLE: u64 = 1 << 1;
    pub const USER: u64 = 1 << 2;
    pub const WRITE_THROUGH: u64 = 1 << 3;
    pub const CACHE_DISABLE: u64 = 1 << 4;
    pub const ACCESSED: u64 = 1 << 5;
    pub const DIRTY: u64 = 1 << 6;
    pub const HUGE_PAGE: u64 = 1 << 7;
    pub const GLOBAL: u64 = 1 << 8;
    pub const NX: u64 = 1 << 63;
    pub const ADDRESS_MASK: u64 = 0x000F_FFFF_FFFF_F000;
}

pub struct Vmm {
    pml4_physical: u64,
    hhdm_offset: u64,
}

unsafe impl Send for Vmm {}

impl Vmm {
    const fn new() -> Self {
        Self {
            pml4_physical: 0,
            hhdm_offset: 0,
        }
    }

    pub fn initialize(&mut self) {
        let hhdm = HHDM.response().expect("HHDM response missing");
        self.hhdm_offset = hhdm.offset;
        self.pml4_physical = read_cr3() & flags::ADDRESS_MASK;
    }

    fn get_pte(&self, virt: u64, create: bool) -> Option<*mut u64> {
        let mut table = (self.pml4_physical + self.hhdm_offset) as *mut u64;

        for level in (1..=3).rev() {
            let index = ((virt >> (12 + 9 * level)) & 0x1FF) as usize;

            unsafe {
                let entry = table.add(index);
                if (*entry & flags::PRESENT) == 0 {
                    if !create {
                        return None;
                    }

                    let frame = PMM.lock().allocate_frame()?;
                    let page = (frame + self.hhdm_offset) as *mut u8;
                    core::ptr::write_bytes(page, 0, 4096);

                    *entry = frame | flags::PRESENT | flags::WRITABLE;
                }
                table = ((*entry & flags::ADDRESS_MASK) + self.hhdm_offset) as *mut u64;
            }
        }

        let index = ((virt >> 12) & 0x1FF) as usize;
        Some(unsafe { table.add(index) })
    }

    pub fn map(&self, virt: u64, phys: u64, page_flags: u64) -> bool {
        match self.get_pte(virt, true) {
            Some(pte) => unsafe {
                *pte = (phys & flags::ADDRESS_MASK) | page_flags | flags::PRESENT;
                true
            },
            None => false,
        }
    }

    pub fn unmap(&self, virt: u64) {
        if let Some(pte) = self.get_pte(virt, false) {
            unsafe { *pte = 0 };
            invlpg(virt);
        }
    }

    pub fn virtual_to_physical(&self, virt: u64) -> Option<u64> {
        let pte = self.get_pte(virt, false)?;
        unsafe {
            if (*pte & flags::PRESENT) == 0 {
                return None;
            }
            Some((*pte & flags::ADDRESS_MASK) | (virt & 0xFFF))
        }
    }

    pub fn physical_to_virtual(&self, phys: u64) -> u64 {
        phys + self.hhdm_offset
    }

    pub fn activate(&self) {
        write_cr3(self.pml4_physical);
    }
}

pub static VMM: Mutex<Vmm> = Mutex::new(Vmm::new());
